src/java/org/apache/cassandra/service/MigrationCoordinator.java

```diff
@@ -0,0 +1,512 @@
+        }
+        return task;
+    }
+
+    private static Future<?> submitToMigrationIfNotShutdown(Runnable task)
+    {
+        LocalAwareExecutorService stage = StageManager.getStage(Stage.MIGRATION);
+
+        if (stage.isShutdown() || stage.isTerminated())
+        {
+            logger.info("Skipped scheduled pulling schema from other nodes: the MIGRATION executor service has been shutdown.");
+            return null;
+        }
+        else
+            return stage.submit(task);
+    }
+
+    @VisibleForTesting
+    protected void mergeSchemaFrom(InetAddress endpoint, Collection<Mutation> mutations) throws IOException
+    {
+        LegacySchemaTables.mergeSchema(mutations);
+    }
+
+    class Callback implements IAsyncCallbackWithFailure<Collection<Mutation>>
+    {
+        final InetAddress endpoint;
+        final VersionInfo info;
+
+        public Callback(InetAddress endpoint, VersionInfo info)
+        {
+            this.endpoint = endpoint;
+            this.info = info;
+        }
+
+        public void onFailure(InetAddress from)
+        {
+            fail();
+        }
+
+        Future<Void> fail()
+        {
+            return pullComplete(endpoint, info, false);
+        }
+
+        public void response(MessageIn<Collection<Mutation>> message)
+        {
+            response(message.payload);
+        }
+
+        Future<Void> response(Collection<Mutation> mutations)
+        {
+            synchronized (info)
+            {
+                if (shouldApplySchemaFor(info))
+                {
+                    try
+                    {
+                        mergeSchemaFrom(endpoint, mutations);
+                    }
+                    catch (Exception e)
+                    {
+                        logger.error(String.format("Unable to merge schema from %s", endpoint), e);
+                        return fail();
+                    }
+                }
+                return pullComplete(endpoint, info, true);
+            }
+        }
+
+        public boolean isLatencyForSnitch()
+        {
+            return false;
+        }
+    }
+
+    private void pullSchema(Callback callback)
+    {
+        if (!isAlive(callback.endpoint))
+        {
+            noSpamLogger.warn("Can't send schema pull request: node {} is down.", callback.endpoint);
+            callback.fail();
+            return;
+        }
+
+        // There is a chance that quite some time could have passed between now and the MM#maybeScheduleSchemaPull(),
+        // potentially enough for the endpoint node to restart - which is an issue if it does restart upgraded, with
+        // a higher major.
+        if (!shouldPullFromEndpoint(callback.endpoint))
+        {
+            logger.info("Skipped sending a migration request: node {} has a higher major version now.", callback.endpoint);
+            callback.fail();
+            return;
+        }
```

Comment:
This check is similar to #95. If we want to include #95, the check should probably go here.